### PR TITLE
Handle errors in finding binary path

### LINKED ISSUE #77: Constant "attempt to concatenate a nil value error"
With the latest version of tabnine-nvim I get the following error message whenever I open Neovim:

```
Error executing vim.schedule lua callback: ...ocal/share/nvim/lazy/tabnine-nvim/lua/tabnine/binary.lua:49: attempt to concatenate a nil value
stack traceback:
        ...ocal/share/nvim/lazy/tabnine-nvim/lua/tabnine/binary.lua:49: in function 'binary_path'
        ...ocal/share/nvim/lazy/tabnine-nvim/lua/tabnine/binary.lua:65: in function 'start'
        ...ocal/share/nvim/lazy/tabnine-nvim/lua/tabnine/binary.lua:114: in function 'request'
        ...ocal/share/nvim/lazy/tabnine-nvim/lua/tabnine/status.lua:19: in function 'cb'
        vim/_editor.lua:263: in function <vim/_editor.lua:262>
```

Worse is whenever I dismiss this error it comes back up a few seconds later.

### PR DESCRIPTION
**Pull request by aarondill**
This avoids constant nil errors (fixes #77), and quits early if the binary can't be found.
Notifications can easily be added in future commits if so desired to warn the user about what is wrong.

to @amirbilu and @vendion, should I add these notifications, or is a silent fail preferable? or, should we write to `config.log_file_path` if it's defined?

### CODE REVIEW

**Comment by vendion** [comment]
I would vote for either silent fail or logging, as a notification would be annoying as well (although shouldn't negatively affect Neovim's usability). I don't need to be constantly reminded that Tabnine doesn't support all the platforms that I use, and would more than likely make me consider canceling my pro subscription.

**Comment by aarondill** [comment]
> I would vote for either silent fail or logging, as a notification would be annoying as well (although shouldn't negatively affect Neovim's usability). I don't need to be constantly reminded that Tabnine doesn't support all the platforms that I use, and would more than likely make me consider canceling my pro subscription.

I could understand not wanting a notification on unsupported hardware (though that could be solved in `lazy.nvim` btw), but should we notify if the binary is not there? that means either the build script failed and needs to be rerun, or else the user (accidentally?) deleted the binary folder w/o rerunning the script.

**Comment by amirbilu** [comment]
guys what's the status here?

**Comment by aarondill** [comment]
We were waiting on your input on the notifications, I believe, but I just added them for missing files, *if* the arch is in our list of supported ones. So this should now be ready to merge.

**Comment by aarondill** [comment]
@amirbilu any update on getting a review here?

File: lua/tabnine/binary.lua [hunk 1]
```diff
@@ -36,6 +38,11 @@ local function binary_name()
 end
 
 local function binary_path()
+	local machine = arch_and_platform()
```
Comment (amirbilu) [inline lua/tabnine/binary.lua:41]: What do you say about rewriting the build script to lua? and have it run on initialization

Comment (aarondill) [reply lua/tabnine/binary.lua:41]: I could do it, but we would end up forking to a shell for `curl`, `unzip` and `chmod` (and their windows equivalents) because there’s no way to do those in lua without external libraries. If thats what you want, I’ll start another PR and do that, but that seems out of scope of this PR

Comment (amirbilu) [reply lua/tabnine/binary.lua:41]: I get it. I believe it would make this a bit easier to manage... I didn't do it to begin with since it requires extra effort which I didn't have at the time.

As for this PR. I would expect `binary_path()` to throw an error in case it fails to find the binary for any case. The current implementation is too scattered in my opinion

Comment (amirbilu) [reply lua/tabnine/binary.lua:41]: yes, i was aiming to use `pcall` & yield the error in one place. does not make sense to you?

Comment (aarondill) [reply lua/tabnine/binary.lua:41]: Sorry, I never saw this reply.

It does not make sense to me to use pcall higher in the call stack when we can use nil and check for it instead.

In my past experience using pcall to detect errors is unwieldy. Error() is kinda like JavaScript's throw, in where you can technically pass any data type, but *most* people use X (strings and Errors, respectively). This makes it hard to *properly* handle errors, and even if strings are properly passed and determined, they are hard to parse due to the possibility that any operation or function call errored and it climbed the call stack. Additionally the inclusion of a call stack in the error string means that we may not even be able to reliably parse the error without a complicated pattern, which would be very prone to typos and bugs.


I would much prefer to keep this limited to the one area, where we can easily check for nil (and an lsp can warn you if you forget).



Furthermore, throwing an error is exactly the problem that caused this PR. @vendion is using unsupported hardware and since this function is called (and errors) frequently, he receives constant error messages (on cursorhold). This pr intends to fix that by detecting misconfigurations and unsupported hardware and warning once for configuration, and being silent for unsupported hardware (per @vendion's request)

Comment (aarondill) [reply lua/tabnine/binary.lua:41]: @amirbilu do you have feedback on this response?
If you think it's a better approach, I can change to using `pcall`, but this will likely make debugging and detecting issues harder in the future. I personally think it's a better idea to return an explicit not-supported value (`nil` in this case), and handle it directly at the call site.
